File: ios/Detonator/Sources/Detonator/Element/ActivityIndicatorElement.swift
import UIKit

class ActivityIndicatorElement: Element {
    override public func decodeAttributes() -> ActivityIndicatorAttributes? {
        return super.decodeAttributes()
    }
    
    override public func createView() -> UIActivityIndicatorView {
        let view = UIActivityIndicatorView()
        
        view.startAnimating()
        
        return view
    }
    
    override public func patchView() {
        let view = view as! UIActivityIndicatorView
        
        let attributes = attributes as! ActivityIndicatorAttributes
        let prevAttributes = prevAttributes as! ActivityIndicatorAttributes?
        
        let size = attributes.size
        let prevSize = prevAttributes?.size
        
        let patchSizeBool = forcePatch || size != prevSize
        
        if patchSizeBool {
            switch size {
            case "medium":
                view.style = .medium
                
                break
                
            case "large":
                view.style = .large
                
                break
                
            default:
                view.style = .medium
                
                break
            }
        }
    }
    
    override func patchColor(color: StyleColor?) {
        let view = view as! UIActivityIndicatorView
        
        view.color = color != nil ? ColorHelper.parseColor(color: color!) : nil
    }

    class ActivityIndicatorAttributes: Attributes {
        var size: String?

        required init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            
            size = try container.decodeIfPresent(String.self, forKey: .size)
            
            try super.init(from: decoder)
        }
        
        private enum CodingKeys: String, CodingKey {
            case size
        }
    }
}

File: android/Detonator/src/main/java/com/iconshot/detonator/element/ActivityIndicatorElement.java
package com.iconshot.detonator.element;

import android.content.res.ColorStateList;
import android.widget.ProgressBar;

import com.iconshot.detonator.Detonator;
import com.iconshot.detonator.helpers.ColorHelper;
import com.iconshot.detonator.helpers.CompareHelper;
import com.iconshot.detonator.layout.ViewLayout.LayoutParams;

public class ActivityIndicatorElement extends Element<ProgressBar, ActivityIndicatorElement.Attributes> {
    public ActivityIndicatorElement(Detonator detonator) {
        super(detonator);
    }

    @Override
    protected Class<Attributes> getAttributesClass() {
        return Attributes.class;
    }

    protected ProgressBar createView() {
        return new ProgressBar(detonator.context);
    }

    @Override
    protected void patchView() {
        LayoutParams layoutParams = (LayoutParams) view.getLayoutParams();

        String size = attributes.size;

        String prevSize = prevAttributes != null ? prevAttributes.size : null;

        boolean patchSize = forcePatch || !CompareHelper.compareObjects(size, prevSize);

        if (patchSize) {
            String tmpSize = size != null ? size : "medium";

            int dp;

            switch (tmpSize) {
                default:
                case "medium": {
                    dp = 20;

                    break;
                }

                case "large": {
                    dp = 37;

                    break;
                }
            }

            int px = dpToPx(dp);

            layoutParams.width = px;
            layoutParams.height = px;
        }
    }

    @Override
    protected void patchColor(Object color) {
        Integer tmpColor = ColorHelper.parseColor(color);

        view.setIndeterminateTintList(tmpColor != null ? ColorStateList.valueOf(tmpColor) : null);
    }

    public static class Attributes extends Element.Attributes {
        String size;
    }
}
